pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,351 @@
+func (r *Reporter) scanAppConfigs(ctx context.Context) error {
+	if len(r.gitRepos) == 0 {
+		r.logger.Info("no repositories were configured for this piped")
+		return nil
+	}
+
+	// Make all repos up-to-date.
+	for repoID, repo := range r.gitRepos {
+		if err := repo.Pull(ctx, repo.GetClonedBranch()); err != nil {
+			r.logger.Error("failed to update repo to latest",
+				zap.String("repo-id", repoID),
+				zap.Error(err),
+			)
+			return err
+		}
+	}
+
+	// Create a map to determine from GitPath if the application is registered.
+	apps := r.applicationLister.List()
+	registeredAppPaths := make(map[string]struct{}, len(apps))
+	for _, app := range apps {
+		key := makeGitPathKey(app.GitPath.Repo.Id, filepath.Join(app.GitPath.Path, app.GitPath.ConfigFilename))
+		registeredAppPaths[key] = struct{}{}
+	}
+
+	if err := r.updateRegisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update registered applications: %w", err)
+	}
+	if err := r.updateUnregisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update unregistered applications: %w", err)
+	}
+
+	return nil
+}
+
+// updateRegisteredApps sends application configurations that have changed since the last time to the control-plane.
+func (r *Reporter) updateRegisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) (err error) {
+	apps := make([]*model.ApplicationInfo, 0)
+	headCommits := make(map[string]string, len(r.gitRepos))
+	for repoID, repo := range r.gitRepos {
+		var headCommit git.Commit
+		headCommit, err = repo.GetLatestCommit(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get the latest commit of %s: %w", repoID, err)
+		}
+		// Skip if the head commit is already scanned.
+		lastScannedCommit, ok := r.lastScannedCommits[repoID]
+		if ok && headCommit.Hash == lastScannedCommit {
+			continue
+		}
+		var as []*model.ApplicationInfo
+		as, err = r.findRegisteredApps(ctx, repoID, repo, lastScannedCommit, headCommit.Hash, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		apps = append(apps, as...)
+		headCommits[repoID] = headCommit.Hash
+	}
+	defer func() {
+		if err == nil {
+			for repoID, hash := range headCommits {
+				r.lastScannedCommits[repoID] = hash
+			}
+		}
+	}()
+	if len(apps) == 0 {
+		return
+	}
+
+	_, err = r.apiClient.UpdateApplicationConfigurations(
+		ctx,
+		&pipedservice.UpdateApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to update application configurations: %w", err)
+	}
+	return
+}
+
+// findRegisteredApps finds out registered application info in the given git repository.
+func (r *Reporter) findRegisteredApps(ctx context.Context, repoID string, repo gitRepo, lastScannedCommit, headCommitHash string, registeredAppPaths map[string]struct{}) ([]*model.ApplicationInfo, error) {
+	files, err := repo.ChangedFiles(ctx, lastScannedCommit, headCommitHash)
```

Comment:
What happens when Piped has just been started and this `lastScannedCommit` was empty?

Reply:
Good catch! I assumed it will be like `git diff --name-only  <headCommitHash>` when it was empty. But Go's exec package doesn't depend on any shell, and either way, that is incorrect.
Let me fix it.

Reply:
👍